crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -1616,18 +1636,18 @@ pub struct NopFetch<State> {
 }
 
 // SAFETY: NopFetch doesnt access anything
-unsafe impl<'w, State: FetchState> Fetch<'w> for NopFetch<State> {
+unsafe impl<'w, F: Fetch<'w>> Fetch<'w> for NopFetch<F> {
     type Item = ();
-    type State = State;
+    type State = F::State;
 
-    const IS_DENSE: bool = true;
+    const IS_DENSE: bool = F::IS_DENSE;
```

Comment:
Given that this doesn't access anything, why can't it be considered dense?

Reply:
Probably shouldn't have changed this in this PR and should have done it separately my bad.

I changed this because I think its _probably_ wrong with `&SparseComponent` nop-ified (although i also didnt actually check... :grimacing:) I think a test case would be a world where we have two entities with components `[Dense, Sparse]` and `[Sparse]` then calling `Query<&Sparse>::is_empty` would return false incorrectly? I realise writing tihs out that `NopWorldQuery` is perhaps analagous to a `Satisfied<Q: WorldQuery>` which might give a better intuition for why `NopWorldQuery<&Sparse>` isnt dense even though it access no data (similarly `With<Sparse>` isnt dense and it accesses no data) 

Just tested: if i set `IS_DENSE` to `true` then this code panics:
```rust
use bevy_ecs::prelude::*;

fn main() {
    let mut world = World::new();

    #[derive(Component)]
    #[component(storage = "SparseSet")]
    struct Sparse;

    #[derive(Component)]
    struct Dense;

    world
        .spawn()
        .insert(Sparse)
        .insert(Dense)
        .remove::<Sparse>();

    assert!(world.query::<&Sparse>().is_empty(&world, 0, 0));
}
```
but with it set to `F::IS_DENSE`  it functions correctly